pkg/ottl: add section about design principles (#29424)
open-telemetry/opentelemetry-collector-contrib: pkg/ottl/LANGUAGE.md

The OTTL grammar includes function invocations, Values and Boolean Expressions. These parts all fit into a Statement, which is the basis of execution in the OTTL.

### Design principles

The OTTL and built-in [ottlfuncs](https://github.com/open-telemetry/opentelemetry-collector-contrib/tree/main/pkg/ottl/ottlfuncs) are designed according to these principles:
- OTTL programs operating in separate contexts cannot influence one another -- an OTTL program may have side-effects only within its own execution [context](#contexts).

> - OTTL programs cannot loop forever, except through the use of non built-in functions.

Reviewer: I really like the idea of documenting design principals, but I am not sure I want to commit to these yet.

> OTTL programs operating in separate contexts cannot influence one another -- an OTTL program may have side-effects only within its own execution [context](#contexts).
While the current contexts have restricted cross-cutting between signals, there isn't anything stopping someone from using OTTL with a custom context that does allow this. OTTL is designed to not worry about how the getting and setting is done or what data is being manipulated. I'd prefer to not restrict ourselves with this principal because some day it may be necessary to support more cross-cutting scenarios.

> OTTL programs cannot loop forever, except through the use of non built-in functions.
Can you provide more what you mean by this statement? I disagree with the term `program` as OTTL is not a programming language.

A design principles for OTTL that I do feel confident about documenting are:
1. Context cannot provide paths that reach "lower" in the OTLP heirarchy.  A function in the metric context, for example, can provide access to the DataPointSlice, but not individual datapoints.
2. OTTL Functions should trend towards erroring when dealing with unexpected results or situations.

Author: @TylerHelmuth thanks for your thoughts!

> While the current contexts have restricted cross-cutting between signals, there isn't anything stopping someone from using OTTL with a custom context that does allow this. OTTL is designed to not worry about how the getting and setting is done or what data is being manipulated. I'd prefer to not restrict ourselves with this principal because some day it may be necessary to support more cross-cutting scenarios.

Sorry, I worded this poorly. What I intended was to (roughly) capture this [Starlark design principle](https://github.com/bazelbuild/starlark#design-principles):

- **Hermetic execution**. Execution cannot access the file system, network, system clock. It is safe to execute untrusted code.

Execution should definitely not be able to access the system or network. I don't see any reason to exclude the system clock, and doing so would would exclude the [`Now()`](https://github.com/open-telemetry/opentelemetry-collector-contrib/tree/main/pkg/ottl/ottlfuncs#now) converter.

In addition to that, what I meant that execution of one OTTL program (sorry, need another word) must not be able to leak into the execution of another.

>> OTTL programs cannot loop forever, except through the use of non built-in functions.
>
> Can you provide more what you mean by this statement?

I was trying to capture what [CEL says](https://github.com/google/cel-spec/blob/master/doc/langdef.md#overview):

- **terminating**: CEL programs cannot loop forever;

The point being to guarantee that CEL can be evaluated without taking down the system evaluating it. It doesn't matter so much for single-tenant systems, since you can just say "don't do that." But for multi-tenant systems it will affect others.

It's not quite enough to say that about the grammar -- as long as there are built-in functions, they would be an escape hatch if they're not also covered by this principle. We can't really enforce what a non built-in function does, hence the comment "except through the use of non built-in functions".

> I disagree with the term program as OTTL is not a programming language.

Right, I wasn't sure what to call it, and it felt a little wrong even if "program" doesn't necessarily mean "general purpose program". FWIW, I went looking for inspiration in cuelang, and found a few references to "CUE program" at:
- https://cuelang.org/docs/references/
- https://cuelang.org/docs/references/spec/
- https://cuelang.org/docs/integrations/go/

Having said that, it doesn't appear to be defined anywhere formally :man_shrugging:

Is there an established word to describe a series of OTTL statements?

Reviewer: > Hermetic execution. Execution cannot access the file system, network, system clock. It is safe to execute untrusted code.

I'm not sure we want to restrict functions in this way.  Although we wouldn't include it in the transform processor, OTTL as a framework allows users to create functions that interact with the network or filesystem. Since function inclusion is per-component and a compile-time decision there is no untrusted code - OTTL does not support dynamic function loading or remote code execution.

> In addition to that, what I meant that execution of one OTTL program (sorry, need another word) must not be able to leak into the execution of another.

This I do agree with.  OTTL statements should be completely independent.

> It's not quite enough to say that about the grammar -- as long as there are built-in functions, they would be an escape hatch if they're not also covered by this principle. We can't really enforce what a non built-in function does, hence the comment "except through the use of non built-in functions".

Ok I see what you're saying.  Yes I think it is safe to say we wouldn't create any functions in ottlfuncs that would loop forever.

> Is there an established word to describe a series of OTTL statements?

Not officially.  I refer to them a `statements`.

Reviewer 2: Could we move these design goals to be in the `ottlfuncs` readme? I think they make a lot of sense for the OTTL standard library, which I would expect to be limited in scope and to have good security guarantees. For the language itself, since OTTL statements run directly in the Go VM we can't make any guarantees for what a user-authored function could do in its implementation.

Author: > Could we move these design goals to be in the ottlfuncs readme? I think they make a lot of sense for the OTTL standard library, which I would expect to be limited in scope and to have good security guarantees.

Sounds good, I'll have a go at that.

> For the language itself, since OTTL statements run directly in the Go VM we can't make any guarantees for what a user-authored function could do in its implementation.

I agree that we can't make any guarantees about user-defined functions, but I still think it's important to address the language to cover any future changes. For example https://github.com/open-telemetry/opentelemetry-collector-contrib/issues/29289 has a few options, one of which is to add looping to the language. That's fine because the proposed loop syntax would guarantee termination. (Obviously you know all that, just giving an example to explain my rationale.)

Author: > Since function inclusion is per-component and a compile-time decision there is no untrusted code - OTTL does not support dynamic function loading or remote code execution.

Understood. Should we capture that here?

Author: Added a mention of no dynamic loading/evaluation, and moved some bits to ottlfuncs.

The OTTL grammar includes function invocations, Values and Boolean Expressions. These parts all fit into a Statement, which is the basis of execution in the OTTL.

### Design principles

OTTL is intended as a domain-specific language (DSL) for telemetry mutation and generation,
and is not intended to be used as a general-purpose programming language.

OTTL has been designed to work directly with [pdata](https://github.com/open-telemetry/opentelemetry-collector/tree/main/pdata).

> While OTTL can work with external data formats due to being implemented directly in Go instead of on top of a custom VM,

Reviewer 2: @TylerHelmuth What do you think about this statement in light of https://github.com/open-telemetry/opentelemetry-collector-contrib/issues/25161? Is it okay for us to have contexts that work pretty much exclusively with external data formats even if it's not "expressly intended", or would we contradict ourselves with those contexts?

Reviewer: I think this is too aggressive of a statement.  I want to leave some wiggle room to handle collector situations like #25161.  Currently OTTL's framework is still very pdata agnostic and the places where we do make direct references to pdata, like some getters/setters, aren't limiting how other people use OTTL.

> OTTL has been designed to work directly with [pdata](https://github.com/open-telemetry/opentelemetry-collector/tree/main/pdata).
> While OTTL can work with external data formats due to being implemented directly in Go instead of on top of a custom VM,
> it is not expressly intended to perform operations that don't touch pdata in some way.

Reviewer 2: ```suggestion
OTTL has been designed to work directly with [pdata](https://github.com/open-telemetry/opentelemetry-collector/tree/main/pdata), but can operate on other data formats.
```

> The OTTL is intended as a domain-specific language (DSL) for telemetry mutation and generation,
> and is not intended to be used as a general-purpose programming language.
> 
> The OTTL has been designed to work directly with [pdata](https://github.com/open-telemetry/opentelemetry-collector/tree/main/pdata).
> While the OTTL can work with external data formats due to being implemented directly in Go instead of on top of a custom VM,
> it is not expressly intended to perform operations that don't touch `pdata` in some way.

Reviewer 2: ```suggestion
OTTL is intended as a domain-specific language (DSL) for telemetry mutation and generation,
and is not intended to be used as a general-purpose programming language.

OTTL has been designed to work directly with [pdata](https://github.com/open-telemetry/opentelemetry-collector/tree/main/pdata).
While OTTL can work with external data formats due to being implemented directly in Go instead of on top of a custom VM,
it is not expressly intended to perform operations that don't touch pdata in some way.
```

I'm sure there are inconsistencies elsewhere, but we made an editorial decision to drop the article from OTTL's name. Also, pdata is usually referred to without the literal quotes.
